pkg/arvo/app/notify.hoon

```diff
@@ -0,0 +1,449 @@
+  --
+::
+++  is-client
+  |=  [who=@p entry=provider-entry]
+  ^-  ?
+  (~(has by clients.entry) who)
+::
+++  post-form
+  |=  [=wire url=@t auth=@t params=(list [@t @t])]
+  ^-  card
+  =/  data
+    %+  roll
+      %+  sort  params
+      |=  [[p=@t @t] [q=@t @t]]
+      (aor p q)
+    |=  [[p=@t q=@t] out=_url]
+    (cat 3 out (cat 3 p q))
+  =/  hmac-sig  (hmac-sha1t:hmac:crypto auth data)
+  =/  b64-sig   (en:base64:mimes:html (met 3 hmac-sig) (swp 3 hmac-sig))
+  =/  headers
+    :~  ['X-Twilio-Signature' b64-sig]
+        ['Content-Type' 'application/x-www-form-urlencoded']
+    ==
+  =/  form-data  (build-form-data params)
+  =/  =request:http
+    [%'POST' url headers `[(met 3 form-data) form-data]]
+  [%pass wire %arvo %i %request request *outbound-config:iris]
+::
+++  build-form-data
+  |=  data=(list [@t @t])
+  ^-  @t
+  %+  roll  data
+  |=  [[p=@t q=@t] out=@t]
+  ?:  =(out '')
+    (cat 3 p (cat 3 '=' q))
+  (cat 3 out (cat 3 '&' (cat 3 p (cat 3 '=' q))))
+::
+++  send-notification
+  |=  [entry=provider-entry who=@p =notification]
+  ^-  card
+  =/  params=(list [@t @t])
+    :~  identity+(rsh [3 1] (scot %p who))
+        ship+(rsh [3 1] (scot %p entity.resource.notification))
+        graph+name.resource.notification
+        :-  %node
+        %+  roll  index.notification
+        |=  [in=@ out=@t]
+        (cat 3 (cat 3 out '/') (scot %ud in))
+    ==
+  %:  post-form
+      /send-notification/(scot %uv (sham eny.bowl))
+      notify-endpoint.entry
+      auth-token.entry
+      params
+  ==
+::
+++  register-binding
+  |=  [service=term entry=provider-entry url=@t who=@p address=@t]
+  ^-  card
+  =/  params=(list [@t @t])
+    :~  identity+(rsh [3 1] (scot %p who))
+        bindingtype+'apn'
+        address+address
+        action+'add'
+    ==
+  %:  post-form
+      /register-binding/(scot %p who)/[service]/(scot %uv (sham eny.bowl))
+      binding-endpoint.entry
+      auth-token.entry
+      params
+  ==
+::
+++  remove-binding
+  |=  [service=term entry=provider-entry who=@p url=@t sid=@t]
+  ^-  card
+  =/  params=(list [@t @t])
+    :~  sid+sid
+        action+'remove'
+    ==
+  %:  post-form
+      /remove-binding/(scot %p who)/[service]/(scot %uv (sham eny.bowl))
+      binding-endpoint.entry
+      auth-token.entry
+      params
+  ==
+::
+++  watch-client
+  |=  [who=@p service=term]
+  ^-  card
+  =/  =wire  /notify/(scot %p who)/[service]
+  [%pass wire %agent [who %notify] %watch /notify/[service]]
+::
+++  leave-client
+  |=  [who=@p service=term]
+  ^-  card
+  =/  =wire  /notify/(scot %p who)/[service]
+  [%pass wire %agent [who %notify] %leave ~]
+::
+++  poke-provider
+  |=  [=wire who=@p act=provider-action]
+  ^-  card
+  [%pass wire %agent [who %notify] %poke %notify-provider-action !>(act)]
+::
+++  give-update
+  |=  =update
+  ^-  card
+  =/  paths=(list path)
+    %+  turn  ~(tap by sup.bowl)
+    |=  [duct =ship =path]
+    path
+  [%give %fact paths %notify-update !>(update)]
```

Comment:
Does this unconditionally give every update to every path? Seems questionable, but I haven't looked closely at how you're using it.